### PR TITLE
Rebuilt ChatSession class

### PR DESCRIPTION
**Pull request by philippjbauer**
- Saves with serialized ChatHistory of session
- Only allows use of ChatHistory.Message (instead of raw text)
   for easy post-processing with IHistoryTransform implementation
- Provides History Management methods
- Allows user to regenerate last assistant message

I also rebased the branch on the last master to keep preview in sync

### CODE REVIEW

File: LLama.Examples/Examples/ChatSessionStripRoleName.cs [hunk 1]
```diff
@@ -1,44 +1,44 @@
...
-namespace LLama.Examples.Examples
-{
-    public class ChatSessionStripRoleName
-    {
-        public static async Task Run()
-        {
-            Console.Write("Please input your model path: ");
-            var modelPath = Console.ReadLine();
-            var prompt = File.ReadAllText("Assets/chat-with-bob.txt").Trim();
+// namespace LLama.Examples.Examples
```
Comment (martindevans) [inline LLama.Examples/Examples/ChatSessionStripRoleName.cs:3]: Is this example no longer relevant?

File: LLama.Examples/Examples/ChatSessionWithRoleName.cs [hunk 2]
```diff
@@ -1,44 +1,44 @@
...
-namespace LLama.Examples.Examples
-{
-    public class ChatSessionWithRoleName
-    {
-        public static async Task Run()
-        {
-            Console.Write("Please input your model path: ");
-            var modelPath = Console.ReadLine();
-            var prompt = File.ReadAllText("Assets/chat-with-bob.txt").Trim();
+// namespace LLama.Examples.Examples
```
Comment (martindevans) [inline LLama.Examples/Examples/ChatSessionWithRoleName.cs:3]: Is this example no longer relevant?

File: .github/workflows/compile.yml [hunk 3]
```diff
@@ -33,7 +33,7 @@ jobs:
             defines: '-DLLAMA_AVX2=OFF'
           - build: 'avx512'
             defines: '-DLLAMA_AVX512=ON'
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
[the remark this answers is not in the document]
Comment (philippjbauer) [reply .github/workflows/compile.yml:36]: The preview branch was quite a bit behind master. I don't think anything went wrong with the rebase, it completed without manual intervention.

File: LLama.Examples/Examples/ChatSessionWithRoleName.cs (hunk 2 above)
Comment (philippjbauer) [reply LLama.Examples/Examples/ChatSessionWithRoleName.cs:3]: It needs to be reworked and tested. I kind of forgot about it ... mea culpa, in between Holiday visits of family right now and have been working on this in the wee morning hours.

File: LLama.Examples/Examples/ChatSessionStripRoleName.cs (hunk 1 above)
Comment (philippjbauer) [reply LLama.Examples/Examples/ChatSessionStripRoleName.cs]: Done in d7096547317151b172bd9150203a76b587b62e8a

File: LLama.Examples/Examples/ChatSessionWithRoleName.cs (hunk 2 above)
Comment (philippjbauer) [reply LLama.Examples/Examples/ChatSessionWithRoleName.cs]: Done in d7096547317151b172bd9150203a76b587b62e8a

File: .github/workflows/compile.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (martindevans) [reply .github/workflows/compile.yml:36]: Oh I see, I hadn't noticed this was going into preview, sorry about that.

In that case I guess we should push up an update to preview to bring it up to date, then we can rebase this PR on that and get a cleaner changeset.

[the remark this answers is not in the document]
Comment (philippjbauer) [reply .github/workflows/compile.yml:36]: Not much cleaner. I'll close this PR and reopen one with the changes committed to the master.

[the remark this answers is not in the document]
Comment (martindevans) [reply .github/workflows/compile.yml:36]: Sorry about all this, you ended up getting caught in the setup of the new preview branch. Hopefully won't happen again!

**Pull request closed without merging**